features/cellular/framework/targets/UBLOX/N2XX/UBLOX_N2XX_CellularStack.cpp

```diff
@@ -0,0 +1,282 @@
+
+    a = _at.read_int();
+    b = _at.read_int();
+
+    socket = find_socket(a);
+    if (socket != NULL) {
+        socket->pending_bytes = b;
+        // No debug prints here as they can affect timing
+        // and cause data loss in UARTSerial
+        if (socket->_cb != NULL) {
+            socket->_cb(socket->_data);
+        }
+    }
+}
+
+int UBLOX_N2XX_CellularStack::get_max_socket_count()
+{
+    return UBLOX_MAX_SOCKET;
+}
+
+bool UBLOX_N2XX_CellularStack::is_protocol_supported(nsapi_protocol_t protocol)
+{
+    return (protocol == NSAPI_UDP);
+}
+
+nsapi_error_t UBLOX_N2XX_CellularStack::create_socket_impl(CellularSocket *socket)
+{
+    int sock_id = 0;
+    int localport = socket->localAddress.get_port();
+
+    if(localport == 5683 && localport < 0 && localport > 65535) {
+        return NSAPI_ERROR_NO_SOCKET;
+    }
+
+    _at.lock();
+    _at.cmd_start("AT+NSOCR=\"DGRAM\",17,");
+    _at.write_int(localport);
+    _at.write_int(1);
+    _at.cmd_stop();
+
+    _at.resp_start();
+    sock_id = _at.read_int();
+    _at.resp_stop();
+
+    if ((_at.get_last_error() != NSAPI_ERROR_OK) || (sock_id == -1)) {
+        _at.unlock();
+        return NSAPI_ERROR_NO_SOCKET;
+    }
+    _at.unlock();
+
+    // Check for duplicate socket id delivered by modem
+    for (int i = 0; i < UBLOX_MAX_SOCKET; i++) {
+        CellularSocket *sock = _socket[i];
+        if (sock && sock->created && sock->id == sock_id) {
+            return NSAPI_ERROR_NO_SOCKET;
+        }
+    }
+
+    socket->id = sock_id;
+    socket->created = true;
+
+    return NSAPI_ERROR_OK;
+}
+
+nsapi_size_or_error_t UBLOX_N2XX_CellularStack::socket_sendto_impl(CellularSocket *socket, const SocketAddress &address,
+        const void *data, nsapi_size_t size)
+{
+    int sent_len = 0, len = 0;
+    char *dataStr = (char *) malloc((size * 2) + 1);
+    memset(dataStr, 0, size*2+1);
+
+    char_str_to_hex_str((const char*)data, size, dataStr);
+    len = strlen(dataStr);
+
+    if (socket->proto == NSAPI_UDP) {
+        if (size > UBLOX_MAX_PACKET_SIZE) {
+            return NSAPI_ERROR_PARAMETER;
+        }
+        _at.cmd_start("AT+NSOST=");
+        _at.write_int(socket->id);
+        _at.write_string(address.get_ip_address());
+        _at.write_int(address.get_port());
+        _at.write_int(len/2);
+        _at.write_string(dataStr);
+        _at.cmd_stop();
+
+        _at.resp_start();
+        _at.skip_param(); // skip socket id
+        sent_len = _at.read_int();
+        _at.resp_stop();
+
+        free(dataStr);
+        if ((_at.get_last_error() == NSAPI_ERROR_OK)) {
+            return sent_len;
+        }
+    }
+
+    return _at.get_last_error();
+}
+
+nsapi_size_or_error_t UBLOX_N2XX_CellularStack::socket_recvfrom_impl(CellularSocket *socket, SocketAddress *address,
+        void *buffer, nsapi_size_t size)
+{
+    nsapi_size_or_error_t nsapi_error_size = NSAPI_ERROR_DEVICE_ERROR;
+    bool success = true;
+    nsapi_size_t read_blk = 0, length = size*2;
```

Comment:
Can you assign read_blk to UBLOX_MAX_PACKET_SIZE so line 175 could be removed. Also better to  move it where it is being used

Reply:
It is checking the case, if requested length is greater max length than only receive max length data on 1st iteration and again check length size on 2nd iteration as if it is greater again if not then just receive remaining length.